### PR TITLE
⚡️ Speed up function `random_block_mask_image` by 64%

### PR DESCRIPTION
**Pull request by misrasaurabh1**
### 📄 64% (0.64x) speedup for ***`random_block_mask_image` in `lightly/models/utils.py`***

⏱️ Runtime :   **`9.47 milliseconds`**  **→** **`5.79 milliseconds`** (best of `58` runs)
### 📝 Explanation and details

Here is an optimized version of your code that addresses the major line-profile time sinks. The bottleneck is repeatedly summing masked areas and assigning to slices, as well as repeated random generation and unnecessary type conversion.

### Key Optimizations.

- **NumPy operations** for the masking region are much faster than PyTorch on CPU for this shape/usage (slice assignment and sum). We carry out mask updates in NumPy, then convert once to torch at the end for maximum slice/assignment/accumulation speed.
- **Precompute random values** where possible, avoiding unnecessary casting and function calls inside tight loops.
- **Minimize .sum().item()** with efficient integer bookkeeping.
- **Remove redundant checks.**
- **Minimize device usage** until a final conversion.
- **Use math.floor/ceil for dimension calculations.**

*The function remains identical in terms of results, comments, and signature.*

**Why this is faster:**
- NumPy operations (`np.count_nonzero`, slice assignment) are much more optimized for this kind of mask bitmap than PyTorch on CPU.
- The sum and update of slices is dramatically faster, eliminating the main bottlenecks in profiling.
- Only one round-trip to torch tensor is performed at the end.

**If your use case is expected to be on CUDA:**
If you *always* run on CUDA, consider using torch throughout but batch the sum/update, but for CPU (default) usage, this yields significant speedup.

Let me know if you need a pure-torch or always-CUDA version as well!

✅ **Correctness verification report:**

| Test                        | Status            |
| --------------------------- | ----------------- |
| ⚙️ Existing Unit Tests | ✅ **51 Passed** |
| 🌀 Generated Regression Tests | ✅ **43 Passed** |
| ⏪ Replay Tests | ✅ **25 Passed** |
| 🔎 Concolic Coverage Tests | 🔘 **None Found** |
|📊 Tests Coverage       | 100.0% |
<details>
<summary>⚙️ Existing Unit Tests Details</summary>

```python
- models/test_ModelUtils.py
- test_tracer_py__replay_test_3.py
```

</details>

<details>
<summary>🌀 Generated Regression Tests Details</summary>

```python
from __future__ import annotations

import math
import random
from typing import Optional, Tuple, Union

# imports
import pytest  # used for our unit tests
import torch
from lightly.models.utils import random_block_mask_image
from torch import Tensor

# --------------------------- BASIC TEST CASES ---------------------------

def test_basic_mask_shape_and_type():
    """Test that the mask has the correct shape and dtype."""
    codeflash_output = random_block_mask_image((8, 8), num_masks=10); mask = codeflash_output # 66.1μs -> 34.8μs

def test_basic_mask_num_true_approximate():
    """Test that the number of masked patches is close to num_masks."""
    codeflash_output = random_block_mask_image((8, 8), num_masks=10); mask = codeflash_output # 44.8μs -> 25.9μs
    num_true = mask.sum().item()

def test_basic_mask_full_image():
    """Test that the function can mask the entire image if num_masks == H*W."""
    H, W = 8, 8
    codeflash_output = random_block_mask_image((H, W), num_masks=H*W); mask = codeflash_output # 341μs -> 205μs

def test_basic_mask_device_cpu():
    """Test that the mask is created on CPU by default."""
    codeflash_output = random_block_mask_image((4, 4), num_masks=5); mask = codeflash_output # 68.4μs -> 46.7μs

def test_basic_mask_device_cuda(monkeypatch):
    """Test that the mask is created on CUDA if device is specified."""
    if torch.cuda.is_available():
        codeflash_output = random_block_mask_image((4, 4), num_masks=5, device='cuda'); mask = codeflash_output # 130μs -> 191μs
    else:
        # If CUDA not available, ensure ValueError is raised if forced
        with pytest.raises(RuntimeError):
            random_block_mask_image((4, 4), num_masks=5, device='cuda')

def test_basic_block_aspect_ratio_default():
    """Test that the default max_block_aspect_ratio is 1/min_block_aspect_ratio."""
    codeflash_output = random_block_mask_image((8, 8), num_masks=10, min_block_aspect_ratio=0.5); mask1 = codeflash_output # 103μs -> 51.1μs

def test_basic_max_num_masks_per_block_none():
    """Test that setting max_num_masks_per_block=None works (should default to num_masks)."""
    codeflash_output = random_block_mask_image((8, 8), num_masks=12, max_num_masks_per_block=None); mask = codeflash_output # 103μs -> 80.8μs

# --------------------------- EDGE TEST CASES ---------------------------



def test_edge_max_num_masks_per_block_less_than_min_raises():
    """Test that ValueError is raised if max_num_masks_per_block < min_num_masks_per_block."""
    with pytest.raises(ValueError):
        random_block_mask_image((8, 8), num_masks=10, min_num_masks_per_block=5, max_num_masks_per_block=3)

def test_edge_max_block_aspect_ratio_less_than_min_raises():
    """Test that ValueError is raised if max_block_aspect_ratio < min_block_aspect_ratio."""
    with pytest.raises(ValueError):
        random_block_mask_image((8, 8), num_masks=10, min_block_aspect_ratio=0.8, max_block_aspect_ratio=0.5)



def test_edge_mask_size_n_by_one():
    """Test an Nx1 image."""
    codeflash_output = random_block_mask_image((7, 1), num_masks=4); mask = codeflash_output # 49.3μs -> 55.9μs

def test_edge_mask_more_than_total_patches():
    """Test that num_masks > total patches masks all patches, no error."""
    codeflash_output = random_block_mask_image((3, 3), num_masks=20); mask = codeflash_output # 17.8μs -> 21.9μs

def test_edge_max_attempts_per_block_zero():
    """Test that max_attempts_per_block=0 results in no masking."""
    codeflash_output = random_block_mask_image((8, 8), num_masks=10, max_attempts_per_block=0); mask = codeflash_output # 9.30μs -> 12.6μs

def test_edge_min_block_aspect_ratio_one():
    """Test that min_block_aspect_ratio=1 (square blocks only) works."""
    codeflash_output = random_block_mask_image((8, 8), num_masks=10, min_block_aspect_ratio=1.0); mask = codeflash_output # 135μs -> 77.2μs

def test_edge_max_block_aspect_ratio_large():
    """Test that a large max_block_aspect_ratio is accepted."""
    codeflash_output = random_block_mask_image((8, 8), num_masks=10, max_block_aspect_ratio=10.0); mask = codeflash_output # 41.0μs -> 25.6μs

def test_edge_random_seed_determinism():
    """Test that setting the random seed gives deterministic results."""
    random.seed(123)
    torch.manual_seed(123)
    codeflash_output = random_block_mask_image((8, 8), num_masks=10); mask1 = codeflash_output # 47.1μs -> 26.7μs
    random.seed(123)
    torch.manual_seed(123)
    codeflash_output = random_block_mask_image((8, 8), num_masks=10); mask2 = codeflash_output # 47.1μs -> 26.7μs

# --------------------------- LARGE SCALE TEST CASES ---------------------------

def test_large_scale_mask_128x128():
    """Test masking on a large image (128x128)."""
    H, W = 128, 128
    num_masks = 1000
    codeflash_output = random_block_mask_image((H, W), num_masks=num_masks); mask = codeflash_output # 162μs -> 78.7μs

def test_large_scale_mask_256x256_partial():
    """Test masking on a larger image (256x256) but with a small num_masks."""
    H, W = 256, 256
    num_masks = 500
    codeflash_output = random_block_mask_image((H, W), num_masks=num_masks); mask = codeflash_output # 163μs -> 55.2μs

def test_large_scale_mask_full():
    """Test masking the entire image on a large image (64x128)."""
    H, W = 64, 128
    num_masks = H * W
    codeflash_output = random_block_mask_image((H, W), num_masks=num_masks); mask = codeflash_output # 770μs -> 471μs

def test_large_scale_max_num_masks_per_block():
    """Test with a small max_num_masks_per_block on a large image."""
    codeflash_output = random_block_mask_image((128, 128), num_masks=500, max_num_masks_per_block=5); mask = codeflash_output # 1.88ms -> 511μs

def test_large_scale_min_num_masks_per_block_large():
    """Test with a large min_num_masks_per_block on a large image."""
    codeflash_output = random_block_mask_image((128, 128), num_masks=500, min_num_masks_per_block=100); mask = codeflash_output # 189μs -> 90.3μs

def test_large_scale_many_masks_close_to_total():
    """Test with num_masks just less than total patches."""
    H, W = 32, 32
    num_masks = H * W - 1
    codeflash_output = random_block_mask_image((H, W), num_masks=num_masks); mask = codeflash_output # 289μs -> 178μs

def test_large_scale_max_attempts_per_block_small():
    """Test with max_attempts_per_block=1 on a large image."""
    H, W = 128, 128
    codeflash_output = random_block_mask_image((H, W), num_masks=100, max_attempts_per_block=1); mask = codeflash_output # 74.5μs -> 48.3μs

def test_large_scale_mask_on_non_square_image():
    """Test on a large non-square image."""
    codeflash_output = random_block_mask_image((100, 300), num_masks=1000); mask = codeflash_output # 150μs -> 76.5μs
# codeflash_output is used to check that the output of the original code is the same as that of the optimized code.

from __future__ import annotations

import math
import random
from typing import Optional, Tuple, Union

# imports
import pytest  # used for our unit tests
import torch
from lightly.models.utils import random_block_mask_image
from torch import Tensor

# unit tests

# Helper function to set random seeds for deterministic tests
def set_seed(seed=42):
    random.seed(seed)
    torch.manual_seed(seed)

# --------------------
# 1. BASIC TEST CASES
# --------------------

def test_basic_mask_shape_and_type():
    """Test that the output mask has the correct shape and type for a typical input."""
    set_seed()
    H, W = 8, 8
    codeflash_output = random_block_mask_image((H, W), num_masks=10); mask = codeflash_output # 66.1μs -> 34.8μs

def test_basic_mask_num_masks_approx():
    """Test that the number of masked patches is close to num_masks (may be slightly more due to block rounding)."""
    set_seed()
    H, W = 8, 8
    num_masks = 12
    codeflash_output = random_block_mask_image((H, W), num_masks); mask = codeflash_output # 118μs -> 86.6μs
    masked_count = int(mask.sum().item())

def test_basic_min_max_num_masks_per_block():
    """Test that ValueError is raised if max_num_masks_per_block < min_num_masks_per_block."""
    with pytest.raises(ValueError):
        random_block_mask_image((8, 8), num_masks=10, min_num_masks_per_block=5, max_num_masks_per_block=4)

def test_basic_min_max_aspect_ratio():
    """Test that ValueError is raised if max_block_aspect_ratio < min_block_aspect_ratio."""
    with pytest.raises(ValueError):
        random_block_mask_image((8, 8), num_masks=10, min_block_aspect_ratio=0.8, max_block_aspect_ratio=0.5)

def test_basic_device_cpu():
    """Test that the mask is created on the specified device (CPU)."""
    set_seed()
    codeflash_output = random_block_mask_image((8, 8), num_masks=10, device="cpu"); mask = codeflash_output # 74.5μs -> 47.8μs

# --------------------
# 2. EDGE TEST CASES
# --------------------


def test_mask_all_patches():
    """Test that requesting to mask all patches results in a mask with all True."""
    set_seed()
    H, W = 8, 8
    codeflash_output = random_block_mask_image((H, W), num_masks=H*W); mask = codeflash_output # 394μs -> 226μs

def test_mask_more_than_possible():
    """Test that requesting more masks than possible does not error and masks all patches."""
    set_seed()
    H, W = 8, 8
    codeflash_output = random_block_mask_image((H, W), num_masks=H*W+10); mask = codeflash_output # 277μs -> 162μs

def test_min_block_aspect_ratio_one():
    """Test with min_block_aspect_ratio=1 (squares only)."""
    set_seed()
    codeflash_output = random_block_mask_image((8, 8), num_masks=10, min_block_aspect_ratio=1.0); mask = codeflash_output # 72.3μs -> 46.3μs
    # All blocks should be square, but we can't check directly; at least check no error

def test_max_block_aspect_ratio_large():
    """Test with a very large max_block_aspect_ratio."""
    set_seed()
    codeflash_output = random_block_mask_image((8, 8), num_masks=10, max_block_aspect_ratio=100.0); mask = codeflash_output # 41.4μs -> 26.6μs
    # Should not error


def test_max_attempts_per_block_zero():
    """Test that with max_attempts_per_block=0, no mask is applied."""
    set_seed()
    codeflash_output = random_block_mask_image((8, 8), num_masks=10, max_attempts_per_block=0); mask = codeflash_output # 26.6μs -> 27.3μs

def test_non_square_image():
    """Test with a non-square image size."""
    set_seed()
    codeflash_output = random_block_mask_image((10, 5), num_masks=10); mask = codeflash_output # 102μs -> 58.1μs

def test_small_image_small_block():
    """Test with a very small image and small block sizes."""
    set_seed()
    codeflash_output = random_block_mask_image((2, 2), num_masks=1, min_num_masks_per_block=1, max_num_masks_per_block=1); mask = codeflash_output # 53.0μs -> 28.9μs

def test_max_num_masks_per_block_larger_than_image():
    """Test with max_num_masks_per_block larger than image area."""
    set_seed()
    codeflash_output = random_block_mask_image((4, 4), num_masks=10, max_num_masks_per_block=100); mask = codeflash_output # 88.2μs -> 72.5μs


def test_large_image_reasonable_runtime_and_memory():
    """Test with a large image to check for performance and memory (under 100MB)."""
    set_seed()
    H, W = 256, 256  # 256*256 = 65536 bools = 8KB
    num_masks = 10000
    codeflash_output = random_block_mask_image((H, W), num_masks); mask = codeflash_output # 279μs -> 131μs

def test_large_num_masks():
    """Test with a large number of masks but not exceeding image size."""
    set_seed()
    H, W = 128, 128
    num_masks = H * W // 2
    codeflash_output = random_block_mask_image((H, W), num_masks); mask = codeflash_output # 283μs -> 178μs
    masked = mask.sum().item()

def test_large_min_block_aspect_ratio_range():
    """Test with a wide range of aspect ratios on a large image."""
    set_seed()
    H, W = 200, 200
    codeflash_output = random_block_mask_image((H, W), num_masks=5000, min_block_aspect_ratio=0.01, max_block_aspect_ratio=100); mask = codeflash_output # 285μs -> 146μs

def test_large_min_num_masks_per_block():
    """Test with a large min_num_masks_per_block to ensure only large blocks are made."""
    set_seed()
    H, W = 100, 100
    num_masks = 500
    min_block = 100
    codeflash_output = random_block_mask_image((H, W), num_masks=num_masks, min_num_masks_per_block=min_block); mask = codeflash_output # 200μs -> 93.8μs

def test_large_max_attempts_per_block():
    """Test with a large max_attempts_per_block to ensure function completes."""
    set_seed()
    H, W = 128, 128
    codeflash_output = random_block_mask_image((H, W), num_masks=1000, max_attempts_per_block=100); mask = codeflash_output # 179μs -> 82.9μs

def test_large_image_mask_all():
    """Test masking all patches in a large image."""
    set_seed()
    H, W = 300, 300  # 90,000 patches, 90KB
    codeflash_output = random_block_mask_image((H, W), num_masks=H*W); mask = codeflash_output # 809μs -> 1.11ms
# codeflash_output is used to check that the output of the original code is the same as that of the optimized code.
```

</details>

[Codeflash](https://codeflash.ai)

### CODE REVIEW

**Comment by misrasaurabh1** [comment]
Hi @guarin, I am working with @liopeer who's been reviewing some of these optimizations offline.
